evm/vm/forks/sharding/guess_head_state_manager.py

```diff
@@ -0,0 +1,245 @@
+
+
+async def download_collation(collation_hash):
+    # TODO: need to implemented after p2p part is done
+    logger.debug("Start downloading collation %s", collation_hash)
+    await asyncio.sleep(SIMULATED_COLLATION_DOWNLOADING_TIME)
+    logger.debug("Finished downloading collation %s", collation_hash)
+    collation = collation_hash
+    return collation
+
+
+async def verify_collation(collation, collation_hash):
+    # TODO: to be implemented
+    logger.debug("Verifying collation %s", collation_hash)
+    await asyncio.sleep(SIMULATED_COLLATION_VERIFICATION_TIME)
+    # data = get_from_p2p_network(collation_hash.data_root)
+    # chunks = DATA_SIZE // 32
+    # mtree = [0] * chunks + [data[chunks*32: chunks*32+32] for i in range(chunks)]
+    # for i in range(chunks-1, 0, -1):
+    #     mtree[i] = sha3(mtree[i*2] + mtree[i*2+1])
+    # assert mtree[i] == collation.data_root
+    return True
+
+
+async def fetch_and_verify_collation(collation_hash):
+    """Fetch the collation body and verify the collation
+
+    :return: returns the collation's validity
+    """
+    collation = await download_collation(collation_hash)
+    return await verify_collation(collation, collation_hash)
+
+
+def clean_done_task(tasks):
+    return [task for task in tasks if not task.done()]
+
+
+class GuessHeadStateManager:
+
+    TIME_ONE_STEP = 0.1
+
+    logger = logging.getLogger("evm.chain.sharding.GuessHeadStateManager")
+
+    last_period_fetching_candidate_head = None
+
+    collation_validity_cache = None
+    chain_validity = None
+
+    unfinished_verifying_tasks = None
+    collation_verifying_task = None
+    chain_collations = None
+
+    def __init__(self, vmc, shard_id, shard_tracker, my_address):
+        # a well-setup vmc handler with a shard_tracker in shard `shard_id`
+        self.vmc = vmc
+        self.shard_id = shard_id
+        self.shard_tracker = shard_tracker
+        self.my_address = my_address
+
+        self.last_period_fetching_candidate_head = 0
+
+        # map[collation] -> validity
+        self.collation_validity_cache = {}
+        # map[chain_head] -> validity
+        # this should be able to be updated by the collations' validity
+        self.chain_validity = defaultdict(lambda: True)
+
+        # current tasks
+        self.unfinished_verifying_tasks = []
+        # map[collation_hash] -> task
+        self.collation_verifying_task = {}
+        # map[chain_head] -> list of collations
+        # the collations in the chain
+        self.chain_collations = defaultdict(list)
+
+    # time related
+
+    def get_current_period(self):
+        return self.vmc.web3.eth.blockNumber // self.vmc.config['PERIOD_LENGTH']
+
+    def is_collator_in_period(self, period):
+        collator_address = self.vmc.get_eligible_proposer(
+            self.shard_id,
+            period,
+        )
+        return collator_address == self.my_address
+
+    def is_collator_in_lookahead_periods(self):
```

Comment:
You're right. I will remove the lookahead related logics outside this class.
Actually I need an indicator of "time's up" in this class.
I'm thinking to add a bool member variable in this class, indicating if currently we are running out of time, and check this variable in every loop. If it is set to `True`, which means running out of time, then just break the loop and return the result of `guess_head`. This is a relatively easy way to do "time's up" in `guess_head`, without introducing collator related logics. And is safe because we use async tasks instead of threads.